ratelimiter/ratelimiter.go

```diff
@@ -0,0 +1,152 @@
+package ratelimiter
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/juju/ratelimit"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// Interface describes common ratelimiter methods.
+type Interface interface {
+	Create([]byte, Config) error
+	Remove([]byte, time.Duration) error
+	TakeAvailable([]byte, int64) int64
+	Available([]byte) int64
+}
+
+// Config is a set of options used by rate limiter.
+type Config struct {
+	Interval, Capacity, Quantum uint64
+}
+
+func newBucket(c Config) *ratelimit.Bucket {
+	return ratelimit.NewBucketWithQuantum(time.Duration(c.Interval), int64(c.Capacity), int64(c.Quantum))
+}
+
+// NewPersisted returns instance of rate limiter with persisted black listed records and capacity before peer was removed.
+func NewPersisted(db DBInterface) *PersistedRateLimiter {
+	return &PersistedRateLimiter{
+		db:          db,
+		initialized: map[string]*ratelimit.Bucket{},
+		now:         time.Now,
+	}
+}
+
+// PersistedRateLimiter persists latest capacity and updated config per unique ID.
+type PersistedRateLimiter struct {
+	db DBInterface
+
+	mu          sync.Mutex
+	initialized map[string]*ratelimit.Bucket
+
+	now func() time.Time
+}
+
+func (r *PersistedRateLimiter) blacklist(id []byte, duration time.Duration) error {
+	if duration == 0 {
+		return nil
+	}
+	record := BlacklistRecord{ID: id, Deadline: r.now().Add(duration)}
+	if err := record.Write(r.db); err != nil {
+		return fmt.Errorf("error blacklisting %#x: %v", id, err)
+	}
+	return nil
+}
+
+func (r *PersistedRateLimiter) get(id []byte) (bucket *ratelimit.Bucket, exist bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	bucket, exist = r.initialized[string(id)]
```

Comment:
```go
return r.initialized[string(id)]
```
should work fine.

Reply:
pretty sure that i have tried that. it didn't unwrap. but i will double check

Reply:
tried this snippet 

```
package main

import (
	"fmt"
)

func check(m map[int]int, k int) (int, bool) {
	return m[k]
}

func main() {
	m := map[int]int{}
	fmt.Println(check(m, 0))
}
```

fails with not enough arguments to return